frontend/src/lib/k8s/api/v2/webSocket.ts

```diff
@@ -1,224 +1,262 @@
+        path: data.path,
+        query: data.query || '',
+        userId: data.userId || '',
+        type: 'CLOSE',
+      };
+      this.socket.send(JSON.stringify(closeMsg));
+    }
+  }
+
   /**
-   * Type of websocket data
+   * Handle WebSocket connection close
    */
-  type?: 'json' | 'binary';
+  private handleWebSocketClose(): void {
+    console.log('WebSocket closed, attempting reconnect...');
+    this.socket = null;
+    this.connecting = false;
+    if (this.listeners.size > 0) {
+      setTimeout(() => this.connect(), 1000);
+    }
+  }
+
   /**
-   * Cluster name
+   * Subscribe to WebSocket updates for a specific path
    */
-  cluster?: string;
+  async subscribe(
+    clusterId: string,
+    path: string,
+    query: string,
+    onMessage: (data: any) => void
+  ): Promise<() => void> {
+    const key = `${clusterId}:${path}`;
+    if (!this.listeners.has(key)) {
+      this.listeners.set(key, new Set());
+    }
+    this.listeners.get(key)!.add(onMessage);
+
+    const socket = await this.connect();
+    const userId = getUserIdFromLocalStorage();
+
+    const message: WebSocketMessage = {
+      clusterId,
+      path,
+      query,
+      userId: userId || '',
+      type: 'REQUEST',
+    };
+
+    socket.send(JSON.stringify(message));
+
+    // Return cleanup function
+    return () => this.handleUnsubscribe(key, onMessage, userId, path, query);
+  }
+
   /**
-   * Message callback
+   * Handle unsubscribe cleanup
    */
-  onMessage: (data: T) => void;
-}) {
-  const url = useMemo(() => (enabled ? createUrl() : ''), [enabled]);
-  const connections = useMemo(() => [{ cluster: cluster ?? '', url, onMessage }], [cluster, url]);
-
-  return useWebSockets({
-    connections,
-    protocols,
-    type,
-  });
-}
+  private handleUnsubscribe(
+    key: string,
+    onMessage: (data: any) => void,
+    userId: string | null,
+    path: string,
+    query: string
+  ): void {
+    const listeners = this.listeners.get(key);
+    listeners?.delete(onMessage);
 
-export type WebSocketConnectionRequest<T> = {
-  cluster: string;
-  url: string;
-  onMessage: (data: T) => void;
-};
+    if (listeners?.size === 0) {
+      this.listeners.delete(key);
+      this.completedPaths.delete(key);
+      if (this.socket?.readyState === WebSocket.OPEN) {
+        const closeMsg: WebSocketMessage = {
+          clusterId: key.split(':')[0],
+          path,
+          query,
+          userId: userId || '',
+          type: 'CLOSE',
+        };
+        this.socket.send(JSON.stringify(closeMsg));
+      }
+    }
+
+    if (this.listeners.size === 0 && this.socket) {
+      this.socket.close();
+      this.socket = null;
+    }
+  }
+}
 
 /**
- * Creates or joins mutiple existing WebSocket connections
- *
- * @param url - endpoint URL
- * @param options - WebSocket options
+ * React hook for WebSocket subscription
  */
-export function useWebSockets<T>({
-  connections,
+export function useWebSocket<T extends KubeObjectInterface>({
```

Comment:
we'll still need to have useWebSockets hook for listening to multiple connections